compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -315,6 +315,7 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
             let emit_as_lint = if let Some(def) = def.as_local() {
                 // (Associated) consts only emit a lint, since they might be unused.
                 matches!(tcx.def_kind(def.did.to_def_id()), DefKind::Const | DefKind::AssocConst)
+                    && !matches!(&err.error, InterpError::MachineStop(err) if err.is_hard_err())
```

Comment:
While at it, could you do what I did in https://github.com/rust-lang/rust/pull/86194 and move this logic to an `is_hard_err` method in an `impl InterpError` block?